#' Perform Single-Cell Quality Control for 10x Genomics Data
#'
#' This function performs a series of quality control (QC) steps on single-cell RNA sequencing data 
#' generated by 10x Genomics. It includes reading the data, detecting and removing empty droplets, 
#' calculating per-cell QC metrics, applying dataset-sensitive and custom filters, and generating 
#' filtered datasets.
#'
#' @param x10_file_path Character. Path to the 10x Genomics data directory.
#' @param empty_droplets_fdr_threshold Numeric. False discovery rate (FDR) threshold for identifying 
#' empty droplets. Default is 0.01.
#' @param empty_lower_droplets Logical. Whether to use a lower threshold for detecting empty droplets. 
#' Default is TRUE.
#' @param replace_unfiltered Logical. Whether to replace existing unfiltered data columns in the 
#' SingleCellExperiment object. Default is TRUE.
#' @param save_dataset_sensitive_filtering Logical. Whether to save the dataset-sensitive filtered 
#' dataset instead of the custom filtered dataset. Default is TRUE.
#' @param BPPARAM A BiocParallel parameter object specifying parallelization settings. Default is 
#' `BiocParallel::SerialParam()`.
#'
#' @return A list of `targets` objects representing the QC pipeline steps, including:
#' \itemize{
#'   \item `path`: The input file path.
#'   \item `sce_raw`: The raw SingleCellExperiment object.
#'   \item `empty_droplets`: Results of empty droplet detection.
#'   \item `sce_no_empty_drop`: SingleCellExperiment object after removing empty droplets.
#'   \item `per_cell_qc_metrics`: Per-cell QC metrics.
#'   \item `sce_unfiltered`: SingleCellExperiment object with unfiltered data and QC metrics.
#'   \item `sce_sensitive_filter`: Dataset-sensitive filtered SingleCellExperiment object.
#'   \item `sce_custom_filter`: Custom filtered SingleCellExperiment object.
#'   \item `sce_filtered`: Final filtered SingleCellExperiment object based on the selected filtering strategy.
#' }
#'
#' @details
#' The function performs the following steps:
#' \enumerate{
#'   \item Reads the 10x Genomics data and creates a raw SingleCellExperiment object.
#'   \item Detects empty droplets using the `DropletUtils::emptyDrops` function.
#'   \item Removes empty droplets based on the specified FDR threshold.
#'   \item Calculates per-cell QC metrics, including mitochondrial and ribosomal gene percentages.
#'   \item Applies dataset-sensitive and custom filters to identify low-quality cells.
#'   \item Generates filtered datasets based on the selected filtering strategy.
#' }
#'
#' @examples
#' \dontrun{
#' tar_sc_single_qc(
#'   x10_file_path = "/path/to/10x/data",
#'   empty_droplets_fdr_threshold = 0.01,
#'   empty_lower_droplets = TRUE,
#'   replace_unfiltered = TRUE,
#'   save_dataset_sensitive_filtering = TRUE,
#'   BPPARAM = BiocParallel::SerialParam()
#' )
#' }
#'
#' @importFrom DropletUtils emptyDrops
#' @importFrom scater perCellQCMetrics isOutlier
#' @importFrom purrr map reduce
#' @importFrom tidyr replace_na
#' @importFrom stringr str_which regex
#' @importFrom BiocParallel SerialParam
#' @importFrom SingleCellExperiment colData rowData counts
#' @export

tar_sc_single_qc <- function(
  x10_file_path,
  empty_droplets_fdr_threshold = 0.01,
  empty_lower_droplets = TRUE,
  replace_unfiltered = TRUE,
  save_dataset_sensitive_filtering = TRUE,
  BPPARAM = BiocParallel::SerialParam()
) {

  targets::tar_assert_path(x10_file_path, 'Input path for 10x data was not found.')


  # 1
  read_10x_counts <- function(path) {
    sce_raw <- read10xCounts(path)
    colnames(sce_raw) <- colData(sce_raw)$Barcode
    sce_raw
  }

  # 2
  detect_empty_droplets <- function(sce_raw, empty_lower_droplets, BPPARAM) {
    DropletUtils::emptyDrops(
      m = counts(sce_raw), lower = empty_lower_droplets, BPPARAM = BPPARAM
    )
  }

  # 3
  remove_empty_drop <- function(sce_raw, empty_droplets, empty_droplets_fdr_threshold) {

    stopifnot(
      ncol(sce_raw) > 0
    )
    if (is.null(empty_droplets)) {
      sce_raw$is_empty_fdr <- NA
    } else {
      is_cell_index <- which(empty_droplets$FDR <= empty_droplets_fdr_threshold)
      colnames_orig <- colnames(sce_raw)
      sce_raw <- sce_raw[, is_cell_index]
      sce_raw$is_empty_fdr <- empty_droplets$FDR[is_cell_index]
    }

    sce_raw
  }

  # 4
  cal_per_cell_qc_metrics <- function(sce_no_empty_drop, BPPARAM) {
    mito_genes <- stringr::str_which(rowData(sce_no_empty_drop)[["Symbol"]], stringr::regex("MT-", ignore_case = TRUE))
    ribo_genes <- stringr::str_which(rowData(sce_no_empty_drop)[["Symbol"]], stringr::regex("^RP[SL]", ignore_case = TRUE))

    scater::perCellQCMetrics(
      sce_no_empty_drop,
      subsets = list(mito = mito_genes, ribo = ribo_genes), 
      BPPARAM = ignore(BPPARAM)
    )
  }

  # 5
  create_unfiltered_sce <- function(sce_no_empty_drop, per_cell_qc_metrics, replace) {
    # ### Dataset-sensitive cell filtering ##########################################
    # MAD_THRESHOLD: 3
    # DATASET_SENSITIVE_FILTERS_OPERATOR: "&"
    # ###############################################################################

    # ### Custom cell filtering #####################################################
    # MIN_UMI_CF: 1000
    # MAX_UMI_CF: 50000
    # MIN_FEATURES: 1000
    # MAX_MITO_RATIO: 0.2
    # CUSTOM_FILTERS_OPERATOR: "&"
    # ###############################################################################

    qc_filter <- list(
        qc_lib = scater::isOutlier(per_cell_qc_metrics$total, log = TRUE, nmads = 3, type = "lower"),
        qc_nexprs = scater::isOutlier(per_cell_qc_metrics$detected, nmads = 3, log = TRUE, type = "lower"),
        qc_mito = scater::isOutlier(per_cell_qc_metrics$subsets_mito_percent, nmads = 3, type = "higher")
        # qc_ribo = isOutlier(per_cell_qc_metrics$subsets_ribo_percent, nmads = !!cfg$MAD_THRESHOLD, type = "higher")
      ) %>%
      purrr::map(~ as.logical(.) %>% tidyr::replace_na(replace = FALSE)) %>%
      purrr::reduce(`&`)


    ## -- Custom filters.
    custom_filter <- list(
        low_count = per_cell_qc_metrics$total <= 1000,
        high_count = per_cell_qc_metrics$total >= 50000,
        low_expression = per_cell_qc_metrics$detected <= 1000,
        high_mito = per_cell_qc_metrics$subsets_mito_percent >= 0.2 * 100
        # low_ribo = per_cell_qc_metrics$subsets_ribo_percent <= !!cfg$MIN_RIBO_RATIO * 100
      ) %>%
    purrr::map( ~ as.logical(.) %>% tidyr::replace_na(replace = FALSE)) %>%
    purrr::reduce(`&`)

    ## -- Add filters to sce and create Seurat object.

    df = data.frame(per_cell_qc_metrics, discard_qc = qc_filter, discard_custom = custom_filter)

    stopifnot(
      nrow(df) == ncol(sce_no_empty_drop)
    )

    if (replace) {
      existing_columns <- intersect(colnames(colData(sce_no_empty_drop)), colnames(df))
      colData(sce_no_empty_drop) <- colData(sce_no_empty_drop)[, !colnames(colData(sce_no_empty_drop)) %in% existing_columns]
    }

    colData(sce_no_empty_drop) <- cbind(colData(sce_no_empty_drop), df)
    
    sce_no_empty_drop
  }

  # 6
  make_sensitive_filter <- function(sce_unfiltered) {
    sce_unfiltered[, !sce_unfiltered$discard_qc]
  }

  # 7
  make_custom_filter <- function(sce_unfiltered) {
    sce_unfiltered[, !sce_unfiltered$discard_custom]
  }

  # 8
  after_filter_rowsums <- function(sce_filter) {
    sce_filtered %>% 
      counts() %>% 
      rowSums()
  }

  # 9
  apply_filter <- function(sce_sensitive_filter, sce_custom_filter, save_dataset_sensitive_filtering) {

    min_umi <- 1
    min_ratio_cells <- 0.01

    if (save_dataset_sensitive_filtering) {
      sce_filter <- sce_sensitive_filter
    } else {
      sce_filter <- sce_custom_filter
    }

    num_cells <- min_ratio_cells * ncol(sce_filter)
    sce_gene_filter <- !rowSums(counts(sce_filter) >= min_umi) >= num_cells

    sce_filter[!sce_gene_filter, ]
  }

  list(
    tar_target_raw(path, x10_file_path, format = "file"),
    tar_target_raw(
      sce_raw, 
      read10xCounts(path)
    ),
    tar_target_raw(
      empty_droplets, 
      detect_empty_droplets(sce_raw, empty_lower_droplets, BPPARAM)
    ),
    tar_target_raw(
      sce_no_empty_drop, 
      remove_empty_drop(sce_raw, empty_droplets, empty_droplets_fdr_threshold)
    ),
    tar_target_raw(
      per_cell_qc_metrics, 
      cal_per_cell_qc_metrics(sce_no_empty_drop, BPPARAM)
    ),
    tar_target_raw(
      sce_unfiltered, 
      create_unfiltered_sce(sce_no_empty_drop, per_cell_qc_metrics, replace_unfiltered)
    ),
    tar_target_raw(
      sce_sensitive_filter, 
      make_sensitive_filter(sce_unfiltered)
    ),
    tar_target_raw(
      sce_custom_filter, 
      make_custom_filter(sce_unfiltered)
    ),
    tar_target_raw(
      sce_filtered, 
      apply_filter(sce_sensitive_filter, sce_custom_filter, save_dataset_sensitive_filtering)
    )
  )
}


