#' Perform Single-Cell Quality Control for 10x Genomics Data
#'
#' This function performs a series of quality control (QC) steps on single-cell RNA sequencing data 
#' generated by 10x Genomics. It includes reading the data, detecting and removing empty droplets, 
#' calculating per-cell QC metrics, applying dataset-sensitive and custom filters, and generating 
#' filtered datasets.
#'
#' @param x10_file_path Character. Path to the 10x Genomics data directory.
#' @param empty_droplets_fdr_threshold Numeric. False discovery rate (FDR) threshold for identifying 
#' empty droplets. Default is 0.01.
#' @param empty_lower_droplets Logical. Whether to use a lower threshold for detecting empty droplets. 
#' Default is TRUE.
#' @param replace_unfiltered Logical. Whether to replace existing unfiltered data columns in the 
#' SingleCellExperiment object. Default is TRUE.
#' @param save_dataset_sensitive_filtering Logical. Whether to save the dataset-sensitive filtered 
#' dataset instead of the custom filtered dataset. Default is TRUE.
#' @param BPPARAM A BiocParallel parameter object specifying parallelization settings. Default is 
#' `BiocParallel::SerialParam()`.
#'
#' @return A list of `targets` objects representing the QC pipeline steps, including:
#' \itemize{
#'   \item `path`: The input file path.
#'   \item `sce_raw`: The raw SingleCellExperiment object.
#'   \item `empty_droplets`: Results of empty droplet detection.
#'   \item `sce_no_empty_drop`: SingleCellExperiment object after removing empty droplets.
#'   \item `per_cell_qc_metrics`: Per-cell QC metrics.
#'   \item `sce_unfiltered`: SingleCellExperiment object with unfiltered data and QC metrics.
#'   \item `sce_sensitive_filter`: Dataset-sensitive filtered SingleCellExperiment object.
#'   \item `sce_custom_filter`: Custom filtered SingleCellExperiment object.
#'   \item `sce_filtered`: Final filtered SingleCellExperiment object based on the selected filtering strategy.
#' }
#'
#' @details
#' The function performs the following steps:
#' \enumerate{
#'   \item Reads the 10x Genomics data and creates a raw SingleCellExperiment object.
#'   \item Detects empty droplets using the `DropletUtils::emptyDrops` function.
#'   \item Removes empty droplets based on the specified FDR threshold.
#'   \item Calculates per-cell QC metrics, including mitochondrial and ribosomal gene percentages.
#'   \item Applies dataset-sensitive and custom filters to identify low-quality cells.
#'   \item Generates filtered datasets based on the selected filtering strategy.
#' }
#'
#' @examples
#' \dontrun{
#' tar_sc_single_qc(
#'   x10_file_path = "/path/to/10x/data",
#'   empty_droplets_fdr_threshold = 0.01,
#'   empty_lower_droplets = TRUE,
#'   replace_unfiltered = TRUE,
#'   save_dataset_sensitive_filtering = TRUE,
#'   BPPARAM = BiocParallel::SerialParam()
#' )
#' }
#'
#' @importFrom DropletUtils emptyDrops
#' @importFrom scater perCellQCMetrics isOutlier
#' @importFrom purrr map reduce
#' @importFrom tidyr replace_na
#' @importFrom stringr str_which regex
#' @importFrom BiocParallel SerialParam
#' @importFrom SingleCellExperiment colData rowData counts
#' @export

tar_sc_single_qc <- function(
  name,
  x10_file_path,
  empty_droplets_fdr_threshold = 0.01,
  empty_lower_droplets = TRUE,
  replace_unfiltered = TRUE,
  save_dataset_sensitive_filtering = TRUE,
  BPPARAM = BiocParallel::SerialParam()
) {

  targets::tar_assert_path(x10_file_path, 'Input path for 10x data was not found.')

  list(
    tar_target_raw("path", x10_file_path, format = "file"),
    tar_target_raw(
      name = "sce_raw", 
      command = quote(scTargets::tar_sc_single_qc_step_read_10x_counts)
    ),
    tar_target_raw(
      name = "empty_droplets", 
      command = substitute(
        scTargets::tar_sc_single_qc_step_detect_empty_droplets(sce_raw, empty_lower_droplets, BPPARAM),
        env = list(empty_lower_droplets = empty_lower_droplets, BPPARAM = BPPARAM)
      )
    )
    # tar_target_raw(
    #   "sce_no_empty_drop", 
    #   quote(scTargets::tar_sc_single_qc_step_remove_empty_drop(sce_raw, empty_droplets, empty_droplets_fdr_threshold))
    # ),
    # tar_target_raw(
    #   "per_cell_qc_metrics", 
    #   quote(scTargets::tar_sc_single_qc_step_cal_per_cell_qc_metrics(sce_no_empty_drop, BPPARAM))
    # ),
    # tar_target_raw(
    #   "sce_unfiltered", 
    #   quote(scTargets::tar_sc_single_qc_step_create_unfiltered_sce(sce_no_empty_drop, per_cell_qc_metrics, replace_unfiltered))
    # ),
    # tar_target_raw(
    #   "sce_sensitive_filter", 
    #   quote(scTargets::tar_sc_single_qc_step_make_sensitive_filter(sce_unfiltered))
    # ),
    # tar_target_raw(
    #   "sce_custom_filter", 
    #   quote(scTargets::tar_sc_single_qc_step_make_custom_filter(sce_unfiltered))
    # ),
    # tar_target_raw(
    #   "sce_filtered", 
    #   quote(scTargets::tar_sc_single_qc_step_apply_filter(sce_sensitive_filter, sce_custom_filter, save_dataset_sensitive_filtering))
    # )
  )
}


